### PR TITLE
Fix `_speculative_sampling` implementation

### PR DESCRIPTION
**Pull request by ofirzaf**
# What does this PR do?

Current implementation of `_speculative_sampling` accepts the draft model tokens all the time due to faulty test of the number of matches (`n_matches`). After fixing this issue I found and fixed several more issues in the implementation to reproduce the exact algorithm presented in the paper.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@gante
@echarlaix

### CODE REVIEW

**Review by gante — APPROVED** [review]
Three glaring mistakes. I don't know what was on my mind that day, honestly. And worse, my speculative decoding quality checks didn't even pick this up 👀 

Thank you very much for catching these bugs and, better yet, for correctly fixing them 💛

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -4819,14 +4819,23 @@ def _speculative_sampling(
     p_n_plus_1 = p[:, n_matches, :]
     if n_matches < gamma:
         q_n_plus_1 = q[:, n_matches, :]
-        p_prime = torch.clamp((p_n_plus_1 - q_n_plus_1), min=0).softmax(dim=-1)
+        p_prime = torch.clamp((p_n_plus_1 - q_n_plus_1), min=0)
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:4822]: Context for @amyeroberts: the paper normalizes this vector, as opposed to a softmax 🙃

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -4819,14 +4819,23 @@ def _speculative_sampling(
...
+        p_prime.div_(p_prime.sum())
     else:
         p_prime = p_n_plus_1
     t = torch.multinomial(p_prime, num_samples=1).squeeze(1)[None, :]
 
     # The selected tokens include the matches (if any) plus the next sampled tokens
     if n_matches > 0:
-        valid_tokens = torch.cat((candidate_input_ids[:, -n_matches:], t), dim=-1)
+        valid_tokens = torch.cat(
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:4830]: context for @amyeroberts: we want to pick the candidates on a left-to-right fashion according to `n_matches`, and not right-to-left 🙃

[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/generation/utils.py:4830]: I might be getting mixed up with the indices - but are these equivalent here? 

I think for `new_candidate_input_ids` - we'd do: 

```
valid_tokens = torch.cat((new_candidate_input_ids[:, :n_matches], t), dim=-1)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this! 

+1 on Joao's comment reworking to make more understandable and adding epsilon. 

Could you also add at least one test which fails on main but fixes with this fix? It's fine if it's an integration test checks generated outputs

File: src/transformers/generation/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ofirzaf) [reply src/transformers/generation/utils.py]: Good comment, looks much better like that. Thanks

**Comment by ofirzaf** [comment]
I was thinking about changing `max_new_tokens` inside the `get_candidates` method so we won't generate more tokens than `max_length - 1`:
```python
max_new_tokens = min(int(self.num_assistant_tokens), self.generation_config.max_length - new_cur_len - 1)
```
This will help avoid checking in several places if `n_matches` is too big, it can also cause a problem in `_speculative_sampling`, and prevent unnecessary compute. 

What do you think?

File: src/transformers/generation/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py]: oops, yes, `:n_matches`

File: src/transformers/generation/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ofirzaf) [reply src/transformers/generation/utils.py:4823]: I don't think it is necessary, if `p_n_plus_1 == q_n_plus_1` then you would have accepted this token in the first place because `p_i / q_i == 1` in this case. From what I see the problem is caused when we generate more candidate tokens then `max_length`.

**Comment by gante** [comment]
@ofirzaf good suggestion, the candidate model should indeed be set up such that the resulting candidate sequence has at most a length of `max length - 1`, as the main model will always generate one additional token 👍

File: src/transformers/generation/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:4825]: Ah, I see, you escape when max_new_tokens is 0.

Let me see if I can find a prettier way of encoding that clause :)

[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:4825]: @ofirzaf I see what's happening. In fact, we need several fixes 🤗 Your solution works in most cases. However, there would still be one corner case when the candidate generator generates a sequence ending with EOS token AND the main model accepts that token. In a correct setup, generation would stop there (we shouldn't generate beyond an EOS token), but the current algorithm is asking the main model to resample the last token (instead of selecting the EOS token) in that situation.

I believe the following will work:
1. Update the candidate model's `max_new_tokens` as you described. Remove your other changes in the last two commits;
2. Update `gamma` to `gamma = min(candidate_logits.shape[1], max_matches)`. This ensures we don't accept more candidates than we should, as `max_matches` contains the maximum admissible length discounting the token that will be sampled from the main model;
3. Inside `if last_assistant_token_is_eos and n_matches == candidate_length:`, add `valid_tokens = new_candidate_input_ids[:, :n_matches + 1]` after `n_matches -= 1`. In a nutshell, don't resample from the main model when the last accepted token is an EOS while keeping `n_matches` consistent with the logic to correctly slice optional output tensors such as `past_key_values`;
4. All lines between `n_matches = min(n_matches, max_matches)` and `valid_tokens = t` should be inside an `else:` block -- an `else` that opposes to `if last_assistant_token_is_eos and n_matches == candidate_length:` clause. This ensures 3. works correctly and jumps straight to the `return` when triggered.

Since you've got a keen eye for speculative decoding, I'd like to kindly ask to review these proposed changes 🤗

[the remark this answers is not in the document]
Comment (ofirzaf) [reply src/transformers/generation/utils.py]: Sounds good. Do you still need to do `n_matches -= 1`? Since we do accept the match of the EOS.
https://github.com/huggingface/transformers/blob/190a63a0b9741cf7c2596557f11eb4da6203c04f/src/transformers/generation/utils.py#L4815
<p>
Another thing that I tried to fix which now happens with this solution is the following case:
Suppose in `get_candidates` you get `max_length - new_cur_len - 1 == 0`, in this case you still generate a single candidate (I guess `generate` ignores `max_new_tokens`) and then the target model will get this token as an input, creating new logits for 2 tokens. Then, speculative will reject/accept this token but since `max_matches == 0` we will ignore that and sample a token from the target.<br>
So now the solution is correct but in this corner case you do spend unnecessary compute for logits/tokens you won't use by definition.<br>
In the last commit that I deleted I avoided that by skipping the generation of candidate tokens in case there is no need.
<p>
What do you think?

**Comment by danielkorat** [comment]
> Thanks for fixing this!
> 
> +1 on Joao's comment reworking to make more understandable and adding epsilon.
> 
> Could you also add at least one test which fails on main but fixes with this fix? It's fine if it's an integration test checks generated outputs

I've been trying to verify the correctness of the Speculative Decoding (SD) implementation. To do so, I'm planning to add a test which verifies that the token distributions we get using SD are indeed similar (or very close) to the token distributions when using standard sample based decoding. If I understand correctly, this is what the SD paper guarantees. 

So far, I found that when comparing the token scores of the above methods on a sample input, they are very similar but not identical, perhaps only sue to hardware numerics errors.
Since tokens with very low probability do not affect the sampling outcome, I think the test should verify that the top `k` token scores of both methods are close enough to each other (for some chose `k` like 5-10). Do you think there's a better way?

When I benchmark Assisted Generation vs SD, I get very similar latencies, even though SD should theoretically be faster, due to a more relaxed acceptance criteria. In consequence, I think it's important to verify the correctness of the SD implementation e.g. by comparing outputs like above. 

@gante @ofirzaf 
WDYT?

My results on A100:
```bash
Target: bigcode/starcoder
Assistant: bigcode/tiny_starcoder_py

Data: openai_humaneval, 20 random examples (args.seed=42)

Setup: 
args.seed=42
max_new_tokens=128
output_scores=True

======================================================================================================================================================
Method               Token Latency        Acceptance Rate      Gen. Args
======================================================================================================================================================

sample               45.19ms              ------               {'do_sample': True, 'temperature': 0.2, 'assistant_model': False}
sd                   25.53ms              67.81%               {'do_sample': True, 'temperature': 0.2, 'assistant_model': True}
ag                   24.04ms              70.78%               {'assistant_model': True}
======================================================================================================================================================
```

**Review by amyeroberts — COMMENTED** [review]
@ofirzaf Thanks for iterating - looks great! 

Once a test is added, happy to approve. I'll let you and @gante discuss the best way to do this, including @danielkorat's suggestion. 

The currently failing tests are unrelated to this PR and a known issue - we're currently trying to fid a fix.

**Comment by ofirzaf** [comment]
@amyeroberts I am not sure what kind of test do you expect to see. I wrote the following test which tests `_speculative_sampling` directly with dummy input data. Is this the kind of testing you are looking for?
```python
import unittest    
from transformers.generation.utils import _speculative_sampling
import torch

class Test(unittest.TestCase):
    def test_speculative_sampling(self):
        # assume vocab size 10, input length 5 + 3 generated candidates
        candidate_input_ids = torch.tensor([[8, 0, 3, 9, 8, 1, 4, 5]])  # input tokens
        candidate_logits = torch.tensor([[
            [-10., 10., -10., -10., -10., -10., -10., -10., -10., -10.],  # generated 1
            [-10., -10., -10., -10., 10., -10., -10., -10., -10., -10.],  # generated 4
            [-10., -10., -10., -10., -10., 10., -10., -10., -10., -10.],  # generated 5
            ]])
        candidate_length = 3
        inf = float('inf')
        new_logits = torch.tensor([[
            [-10., 10., -10., -10., -10., -10., -10., -10., -10., -10.],  # accepts 1
            [-10., -10., -10., -10., 10., -10., -10., -10., -10., -10.],  # accepts 4
            [-inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, 10., -inf],  # rejects 5, accepts 8
            [-10., -10., -10., -10., -10., -10., -10., -10., -10., -10.],  # N/A
        ]])
        last_assistant_token_is_eos=False
        max_matches = 5
        validated_tokens, n_matches = _speculative_sampling(
            candidate_input_ids,
            candidate_logits,
            candidate_length,
            new_logits,
            last_assistant_token_is_eos,
            max_matches,
        )
        self.assertTrue(n_matches.item() == 2)
        self.assertTrue(validated_tokens.tolist()[0] == [1, 4, 8])
```

**Comment by ofirzaf** [comment]
@danielkorat we want to test that the probability to sample token $t$ from the target model is the same as the probability to sample $t$ using speculative. Maybe the following test will work, given a prompt and a target model, calculate the distribution for the next token by simply running the model on the prompt and applying `softmax` on the resulted logits. Then, we can generate $n$ guesses for the next token using speculative. If speculative is implemented correctly and $n$ is large enough you should get a distribution that would match the calculated distribution from before. 

Does that make sense? WDYT?

**Comment by danielkorat** [comment]
@ofirzaf Yes we can either do that or compare the logits using `np.allclose()`.
I wonder though if the guarantee holds for the next token only and not for all subsequent tokens.
If it holds only for the next token, we can have test feed the SD algo with standard sampled input_ids one at a time and compare outputs (one at a time) in one of the suggested methods above.
________________________________

File: src/transformers/generation/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py]: We do need that line (`n_matches -= 1`), as some output slicing code (e.g. [here](https://github.com/huggingface/transformers/blob/c662c78c71e693234725d70a6291398b33042444/src/transformers/generation/utils.py#L4435)) assumes the output of the model at a given generation step is `n_matches + 1`.

We should add a comment on this line, though, mentioning why it exists!

[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py]: Re corner cases you've identified: tough one. On one hand, we save some computation on the smaller model. On the other hand, it introduces a few extra if/elses, hurting readability. 

Since the point of speculative decoding is performance, let's fix the corner case as you suggested. One extra request, though: on the added if/elses to handle this corner cases, add a small comment explaining why it's there :)

**Comment by gante** [comment]
@ofirzaf your test suggestion would be a great addition 👍

**Comment by gante** [comment]
@danielkorat If the claim in the paper is correct, then @ofirzaf suggestion should work regardless of the distance of the target token to the end of the prompt. In other words, we should see the same distribution in the 1st generated token, in the 2nd generated token given the 1st generated token, and so on.

My suggestion would be to generate ~10 tokens for a given prompt ~1M times using super small models (e.g. pythia), to confirm it empirically.

File: src/transformers/generation/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ofirzaf) [reply src/transformers/generation/utils.py]: Found a way to avoid an extra if :)

**Comment by ofirzaf** [comment]
@gante @amyeroberts I have pushed all the changes we discussed. Can you please take a look?

**Comment by danielkorat** [comment]
@gante @ofirzaf 
I wrote this code to check our assumptions about the distributions:

```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, set_seed

def top_k(a, k):
    sorted, indices = a.sort(descending=True)
    return sorted[:k], indices[:k]
def get_dist(outputs, num_samples=10**6, norm_factor=1e7):
    probs = outputs.scores[0][0].softmax(dim=-1)
    print("probs:", top_k(probs, k=5))
    next_std_token = torch.multinomial(probs, num_samples=num_samples, replacement=True)
    counts = torch.bincount(next_std_token, minlength=probs.shape[0]).float()
    print("counts:", top_k(counts, k=5))
    norm_counts = counts.float() / norm_factor
    print("softmaxed_counts:", top_k(counts.softmax(dim=-1), k=5)[0])
    print("norm_counts:", top_k(norm_counts, k=5))
    softmaxed_norm_counts = top_k(norm_counts.softmax(dim=-1), k=5)[0]
    print("softmaxed_norm_counts:", softmaxed_norm_counts)
    return softmaxed_norm_counts

checkpoint = "EleutherAI/pythia-1.4b-deduped"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForCausalLM.from_pretrained(checkpoint)
assistant_model = AutoModelForCausalLM.from_pretrained("EleutherAI/pythia-160m-deduped")
gen_kwargs = dict(**tokenizer("In other words, we", return_tensors="pt"),
                  pad_token_id=tokenizer.eos_token_id,
                  max_new_tokens=1,
                  do_sample=True,
                  temperature=0.1,
                  return_dict_in_generate=True,
                  output_scores=True)
print("Standard sampling:")
set_seed(0)
std_dist = get_dist(model.generate(**gen_kwargs))
print("\nSpeculative decoding:")
set_seed(0)
sd_dist = get_dist(model.generate(**gen_kwargs, assistant_model=assistant_model))
print(std_dist == sd_dist)
```

In short, I get the token probs for some input for both decoding methods, then I sample 1M times from this prob, and get the token counts (I always look at top 5 scores in all the following computations). As you can see below, the `counts` are very close to each other but not equal of course. If we apply `softmax` to the counts, they produce an insignificant result of  [1., 0., 0., 0., 0.] (tested other inputs too). So I normalize the counts and then apply `softmax`, and then the output distributions look similar, however, they are still not equal to each other (final output). 

WDYT? Is this the way to go?

Output::
```bash
Standard sampling:
probs: (tensor([5.9927e-01, 3.8301e-01, 1.7703e-02, 8.4421e-06, 1.5724e-06]), tensor([452, 403, 476, 878, 513]))
counts: (tensor([5.9925e+05, 3.8286e+05, 1.7879e+04, 1.7000e+01, 1.0000e+00]), tensor([452, 403, 476, 878, 513]))
softmaxed_counts: tensor([1., 0., 0., 0., 0.])
norm_counts: (tensor([5.9925e-02, 3.8286e-02, 1.7879e-03, 1.7000e-06, 1.0000e-07]), tensor([452, 403, 476, 878, 513]))
softmaxed_norm_counts: tensor([2.1106e-05, 2.0654e-05, 1.9914e-05, 1.9878e-05, 1.9878e-05])

Speculative decoding:
probs: (tensor([5.9925e-01, 3.8304e-01, 1.7706e-02, 8.4404e-06, 1.5720e-06]), tensor([452, 403, 476, 878, 513]))
counts: (tensor([5.9922e+05, 3.8288e+05, 1.7883e+04, 1.7000e+01, 1.0000e+00]), tensor([452, 403, 476, 878, 457]))
softmaxed_counts: tensor([1., 0., 0., 0., 0.])
norm_counts: (tensor([5.9922e-02, 3.8288e-02, 1.7883e-03, 1.7000e-06, 1.0000e-07]), tensor([452, 403, 476, 878, 457]))
softmaxed_norm_counts: tensor([2.1106e-05, 2.0654e-05, 1.9914e-05, 1.9878e-05, 1.9878e-05])
2.110590685333591e-05
2.1105903215357102e-05
tensor([False, False, False, False, False])
```

**Comment by gante** [comment]
@danielkorat there are tiny fluctuations to be expected: doing the forward pass with different shapes (e.g. 1 token at a time vs all at once, 1 row at a time vs batched) will result in slightly different outputs. I've written about it in more detail [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).

Factoring this source of numerical differences, the results do look similar! Thank you for running the experiment 🤗

**Review by gante — APPROVED** [review]
@ofirzaf Thank you for iterating with us 💛 

@amyeroberts I'm happy with the PR, ready for your final check

**Comment by danielkorat** [comment]
> @danielkorat there are tiny fluctuations to be expected: doing the forward pass with different shapes (e.g. 1 token at a time vs all at once, 1 row at a time vs batched) will result in slightly different outputs. I've written about it in more detail [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).
> 
> 
> 
> Factoring in this source of numerical differences, the results do look similar! Thank you for running the experiment 🤗 

@gante My intention is to integrate this code into the repo tests. WDYT?
Should I do this in a separate PR?

**Comment by gante** [comment]
@danielkorat This test requires a lot of compute, so it's not fit for our CI 😅

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -2351,6 +2352,43 @@ def test_top_k_top_p_filtering_with_filter_value(self):
 
         self.assertTrue(torch.allclose(expected_output, output, atol=1e-12))
 
+    def test_speculative_sampling(self):
```
Comment (amyeroberts) [inline tests/generation/test_utils.py:2355]: Nice :D

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this, iterating to make it a mean, clean refactor and adding tests ❤️
